core/src/main/java/hudson/TcpSlaveAgentListener.java

```diff
@@ -200,6 +204,14 @@ public ConnectionHandler(Socket s) {
                 id = iotaGen++;
             }
             setName("TCP agent connection handler #"+id+" with "+s.getRemoteSocketAddress());
+            setUncaughtExceptionHandler((t, e) -> {
+                LOGGER.log(Level.SEVERE, "Uncaught exception in TcpSlaveAgentListener " + t, e);
+                try {
+                    s.close();
+                } catch (IOException e1) {
+                    // try to clean up the socket
```

Comment:
Something is missing here, right?
Have you checked that AgentProtocol gets aborted correctly in this case?